nixos/modules/services/databases/postgresql.nix

```diff
@@ -109,7 +109,34 @@ in
       };
 
       ensureDatabases = mkOption {
-        type = types.listOf types.str;
+        type = with types; listOf (either str (submodule {
+          options = {
+            name = mkOption {
+              type = types.str;
+              description = ''
+                Name of the database to create.
+              '';
+            };
+            lc_collate = mkOption {
```

Comment:
I'm not a fan of these options. I think `ensure` is in general a "bad word" for a declarative tool like NixOS. I think Nix isn't a good tool for this type of mechanism. @aanderse and @flokli and I have talked in the past about supporting this in a way similar to the "users" support (mutableUsers / creation / deletion as needed.)

In particular, I'm uncomfortable with these `lc_collate` and `lc_ctype` options because they are not declarative, and don't _ensure_ these databases _are_ as specified. In other words, this is creating a type of untracked state where the configuration.nix is lying about the system, something I think we should try hard to avoid.

Concretely, these two systems will end up in different states, something NixOS typically tries very hard to do:

```
+ ensureDatabases.foo.lc_collate = "foo";
$ nixos-rebuild switch
- ensureDatabases.foo.lc_collate = "foo";
+ ensureDatabases.foo.lc_collate = "bar";
$ nixos-rebuild switch
```


```
+ ensureDatabases.foo.lc_collate = "bar";
$ nixos-rebuild switch
```

I would prefer to not see this part of this PR merge.

Reply:
First of all, thanks a lot for raising your concerns about those patches!

> I think ensure is in general a "bad word" for a declarative tool like NixOS

Those options were introduced earlier (we have `ensure{Users,Databases}` for quite a while now). Also, I'd like to mention that previously options such as [`initialScript`](https://nixos.org/nixos/options.html#services.postgresql.initialscript) were used that were even more impure.

Or in other words: do you have a suggestion how to manage that kind of state in Nix? (Admittedly I'm fairly unsafe when it comes to state in Nix, I mostly use workarounds like those `ensure*`-options or `initialScript` as I'm currently not aware of a better solution).

> @aanderse and @flokli and I have talked in the past about supporting this in a way similar to the "users" support (mutableUsers / creation / deletion as needed.)

In case this was discussed in IRC - do you have a link to that discussion? :) 

> In particular, I'm uncomfortable with these lc_collate and lc_ctype options because they are not declarative, and don't ensure these databases are as specified. In other words, this is creating a type of untracked state where the configuration.nix is lying about the system, something I think we should try hard to avoid.

As already mentioned, I'm rather uncomfortable when it comes to actual state-management with Nix. May I ask if you have a different suggestion (don't get me wrong - I don't expect you to know "the perfect" solution just because you raised concerns about those patches. I'm just curious as I currently don't know any better fix than the one I filed here).

Reply:
I know `ensurexxx` has been here longer than this PR, and I don't like it any better :P. ensureUsers in particular is Very scary for me. If initialScript could work, I think that would be okay-ish. initialScript is less bad, since it has in its name the fact that it is almost guaranteed to be lies and doesn't attempt to pretend that it is declarative.

Reply:
So your main problem are those `ensure*` options, especially if there are additional options to configure a database or a user as those can be wrongly interpreted as declarative? And that the same issue exists with the GRANT-options for `ensureUsers`?

> If initialScript could work, I think that would be okay-ish

I'm not sure if I can agree here. I still remember the first time I used the postgresql-module a few years ago and it took me a while until I figured out that `initialScript` is only executed if `${cfg.dataDir}/.first_startup` doesn't exist. What I want to say is that this can be misinterpreted as well, but in the end the behavior of both `initialScript` and `ensure*` is currently well-documented and doesn't affect existing setups (at least in the case of matrix-synapse).

If you insist on not enhancing and using `ensure*` here, we can still revert #75294 and fix the issue in the systemd-unit.

Otherwise I personally think that this is a general issue with the way Nix "handles" state and further measures should be discussed in a separate issue.

Reply:
`ensureDatabases.*.initializationOptions.*` would be both more convenient and explicit enough in my opinion.

Reply:
The PR is more about `synapse` than `ensureDatabases`. However the later blocks the PR. How can we resolve this deadlock?

Reply:
@mguentner I basically agree with you (that's why I suggested to merge this and discuss the issues about `ensure*` in a separate issue), however I certainly understand @grahamc's concerns about worsening the situtation in `nixos/postgresql` and I'd like to have this resolved before having those patches in upstream.

We'll probably talk about the entire issue synchronously on IRC and try to find a suitable comproise.

Apart from this I'd probably add `matrix-synapse` at [`1.11`](https://github.com/matrix-org/synapse/releases/tag/v1.11.0) before considering this mergable again.

Reply:
Very important point about `lc_ctype` and `lc_collate` violating the current behaviour of `ensure*`, and being more similar to `initialScript`, though with database creation level instead of database software installation level granularity.
As synapse (may?) will eventually disallow databases with incorrect ctype and collation, don't we need some form of automatic ctype & collation migration anyway? That would allow to have a `lc_ctype` and `lc_collate` option that doesn't lie about the system state and behaves with the same guarantees like all other `ensure*` options.

Reply:
I reverted the patch that used ensure* in the matrix-synapse module, so I consider this resolved for now.

> don't we need some form of automatic ctype & collation migration anyway

IMHO no: as we've seen e.g. throughout this discussion Nix doesn't do a very good job at state management. IMHO the safest approach here would be to ask administrators to manually do the migration.

Please note that I just realized that this only affects synapse setups from the last few months as we configured databases properly before switching to ensure* here (b2d67c08d03679ba1b9a928ac5610ad940846e7f).